dask_cuda/explicit_comms/dataframe/shuffle.py

```diff
@@ -5,201 +5,239 @@
-):
+    eps,
+    myrank,
+    rank_to_out_part_ids: Dict[int, Set[int]],
+    out_part_id_to_dataframe_list: Dict[int, List[DataFrame]],
+    proxify,
+) -> None:
     """Notice, received items are appended to `out_parts_list`"""
-    futures = []
-    for rank, ep in eps.items():
-        if rank in in_nparts:
-            futures.append(ep.read())
 
-    # Notice, since Dask may convert lists to tuples, we convert them back into lists
-    out_parts_list.extend(
-        [[y for y in x] for x in nested_deserialize(await asyncio.gather(*futures))]
+    async def read_msg(rank: int) -> None:
+        msg: Dict[int, DataFrame] = nested_deserialize(await eps[rank].read())
+        for out_part_id, df in msg.items():
+            out_part_id_to_dataframe_list[out_part_id].append(proxify(df))
+
+    await asyncio.gather(
+        *(read_msg(rank) for rank in rank_to_out_part_ids if rank != myrank)
     )
 
 
-def sort_in_parts(
-    in_parts: List[Dict[int, DataFrame]],
-    rank_to_out_part_ids: Dict[int, List[int]],
-    ignore_index: bool,
-    concat_dfs_of_same_output_partition: bool,
-    concat,
-) -> Dict[int, List[List[DataFrame]]]:
-    """Sort the list of grouped dataframes in `in_parts`
-
-    It returns a dict that for each worker-rank specifies the output partitions:
-    '''
-        for each worker:
-            for each output partition:
-                list of dataframes that makes of an output partition
-    '''
-    If `concat_dfs_of_same_output_partition` is True, all the dataframes of an
-    output partition are concatenated.
+def get_proxify(worker: Worker) -> Callable[[T], T]:
+    """Get function to proxify objects"""
+    from dask_cuda.proxify_host_file import ProxifyHostFile
+
+    if isinstance(worker.data, ProxifyHostFile):
+        data = worker.data
+        return lambda x: data.manager.proxify(x)[0]
+    return lambda x: x  # no-op
+
+
+def compute_map_index(df: Any, column_names, npartitions) -> Series:
+    """Return a Series that maps each row `df` to a partition ID"""
+    if column_names[0] == "_partitions":
+        ind = df[column_names[0]]
+    else:
+        ind = hash_object_dispatch(
+            df[column_names] if column_names else df, index=False
+        )
+    typ = numpy.min_scalar_type(npartitions * 2)
+    return (ind % npartitions).astype(typ, copy=False)
+
+
+def single_shuffle_group(
+    df: DataFrame, column_names, npartitions, ignore_index
+) -> Dict[int, DataFrame]:
+    """Split dataframe such that each partition hashes to the same"""
+    map_index = compute_map_index(df, column_names, npartitions)
+    return group_split_dispatch(df, map_index, npartitions, ignore_index=ignore_index)
+
+
+def multi_shuffle_group(
+    df_meta: DataFrame,
+    dfs: Dict[str, DataFrame],
+    column_names,
+    npartitions,
+    ignore_index,
+    proxify,
+) -> Dict[int, DataFrame]:
+    """Split multiple dataframes such that each partition hashes to the same
```

Comment:
"hashes to the same" needs a "what" (as in what does it hash the same as).

Reply:
Added docs in feac8d5f56887cf7eff5591d1ea3821a2038c24a